src/mon/commands/Command.h

```diff
@@ -0,0 +1,330 @@
+    service(_svc),
+    cct(_cct)
+  { }
+
+  virtual ~Common() { }
+
+ private:
+  Common() = delete;
+  void wait_for_finished_proposal(MonOpRequestRef op, Context *ctx) {
+    service->wait_for_finished_proposal(op, ctx);
+  }
+
+ protected:
+  Monitor *mon;
+  Service *service;
+  CephContext *cct;
+
+  version_t get_last_committed() {
+    return service->get_last_committed();
+  }
+
+ public:
+  bool reply_with_data(MonOpRequestRef op,
+		       int retcode,
+		       stringstream &ss,
+		       bufferlist &retdata,
+		       version_t version) {
+    string retstr;
+    getline(ss, retstr);
+
+    bool ret = true;
+
+    if (retcode < 0 && retstr.length() == 0) {
+      retstr = cpp_strerror(retcode);
+      ret = false;
+    }
+    mon->reply_command(op, retcode, retstr, retdata, version);
+    return ret;
+  }
+
+  virtual bool reply(
+      MonOpRequestRef op,
+      int retcode,
+      stringstream &ss,
+      version_t version) {
+    bufferlist retdata;
+    return reply_with_data(op, retcode, ss, retdata, version);
+  }
+
+  bool update(MonOpRequestRef op,
+	      stringstream &ss,
+	      version_t version) {
+
+    string retstr;
+    getline(ss, retstr);
+
+    wait_for_finished_proposal(op,
+	new Monitor::C_Command(mon, op, 0, retstr, version));
+    return true;
+  }
+
+  bool wait_retry(MonOpRequestRef op) {
+    wait_for_finished_proposal(op,
+	new Monitor::C_RetryMessage(mon, op));
+    return true;
+  }
+};
+
+
+/* Command
+ *
+ * Define a command. It will be templated with
+ *  T	    = a paxos service
+ *  Stable  = the typename of the paxos service's stable map
+ *  Pending = the typename of the paxos service's pending map
+ *
+ * Depending on a paxos service's specificities, Stable and Pending may be
+ * the same.
+ *
+ * This class doesn't use these types for much more than defining the prepare
+ * and preprocess functions to be called by derived classes, but still
+ * requires them to be defined to properly abstract command definitions.
+ */
+
+typedef std::shared_ptr<Formatter> FormatterRef;
+
+template<typename Service, typename ...Args>
+struct Command : public Common<Service>
+{
+
+  CephContext *cct;
+
+  explicit Command(
+      Monitor *_mon,
+      Service *_service,
+      CephContext *_cct) :
+    Common<Service>(_mon, _service, _cct)
+  { 
+    cct = this->cct;
+  }
+
+  
+  virtual ~Command() { };
+
+  virtual bool preprocess(
+      MonOpRequestRef op,
+      const cmdmap_t &cmdmap,
+      Args ...args) {
+
+    stringstream ss;
+    bufferlist rdata;
+
+    string format;
+    cmd_getval(cct, cmdmap, "format", format, string("plain"));
+    FormatterRef f(Formatter::create(format));
+
+    string prefix;
+    cmd_getval(cct, cmdmap, "prefix", prefix);
+
+    return do_preprocess(op, prefix, cmdmap, ss, rdata, f, args...);
+  }
+
+  virtual bool prepare(
+      MonOpRequestRef op,
+      const cmdmap_t &cmdmap,
+      Args ...args) {
+
+    op->mark_osdmon_event(__func__);
+    stringstream ss;
```

Comment:
It's fine for now, but a lot of the resource-gathering in here seems basically identical-- might be nice to consolidate it once the patterns are clear.